### PR TITLE
Add the K framework language

### PR DESCRIPTION
**Pull request by ayberkt**
Hi!

K is a language for defining programming languages as well as type systems and other formal analysis tools. It was first initiated by Grigore Rosu and is currently developed by the FSL group at the University of Illinois at Urbana-Champaign and the FMSE group at Alexandru Ioan Cuza University in Iasi, Romania. You can visit [the K Framework website](http://www.kframework.org/index.php/Main_Page) for a more in-depth explanation.

Both the developers and users of K have been using GitHub to host K code for a while though GitHub linguist cannot identify it correctly. As the common usage is now exceeding a certain threshold, we thought we should do something about this.

As requested in `CONTRIBUTING.md` [here is the search](https://github.com/search?utf8=%E2%9C%93&q=semantics+extension%3Ak&type=Code&ref=advsearch&l=&l=) showing the common usage of the K Framework.

## Some problems

This is the first time I am contributing to `linguist` so I would appreciate some help by the maintainers. The first issue is that running `./script/set-language-ids --update` did not work for me, resulting in the error:
```
/Users/ayberkt/.rvm/rubies/ruby-2.4.0/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- bundler/setup (LoadError)
        from /Users/ayberkt/.rvm/rubies/ruby-2.4.0/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from ./script/set-language-ids:3:in `<main>'
```
I am not at all experienced with neither Ruby nor RVM so I could not figure out what's going wrong. If this script is just to update the language ids in `languages.yml`, I would appreciate if one of the maintainers could run it for me—if they don't mind doing so; I believe this will be the easiest solution.

Also if I recall correctly, running `script/add-grammar` resulted in some warnings/errors though it seemed to correctly do what it was supposed to do. I would appreciate as well if someone could check whether it worked correctly.

Other than these, all steps listed in `CONTRIBUTING.md` have been completed.

Looking forward to seeing K Framework on GitHub linguist!

/cc: @msaxena2 @lucaspena

### CODE REVIEW

**Comment by ayberkt** [comment]
Any chance we can get an update on this?

/cc: @msaxena2 @lucaspena

**Comment by pchaigno** [comment]
> I shared this same concern in #3638 and I'm not sure how we've addressed this in the past. @larsbrinkhoff, @Alhadis, @pchaigno do you guys know?

In the past, we've postponed these kind of pull request (`.dat` file extension for instance) until we have a `not-a-language` option.

> I'm tempted to go the route of using a heuristic which falls back to "Text". This approach can be used for #3638 too.

So we wouldn't use the Bayesian classifier for `.k` files? Unless there's a simple way to identify K files, I'm afraid a heuristic could only catch common cases. Would that be good enough?

**Comment by lildude** [comment]
> In the past, we've postponed these kind of pull request (`.dat` file extension for instance) until we have a `not-a-language` option.

Aaaah, yes. I remember seeing discussions about that now you point it out.

> So we wouldn't use the Bayesian classifier for .k files? Unless there's a simple way to identify K files, I'm afraid a heuristic could only catch common cases. Would that be good enough?

Hmmm, good point. Probably not. It might work in this case, but thinking about it, it'll probably lead to bad behaviour in the long run.

**Comment by pchaigno** [comment]
We should really add a default choice. Unfortunately, I'm not even sure the current classifier is a good fit for that... Can we compute a confidence for its results?

**Comment by ayberkt** [comment]
Thanks for the help @lildude! Here is a summary of what I did to address the issues you have mentioned.

1. I took a look at the issue with the `grammars.yml` file and it turns out that `script/add-grammar` had not been properly run but still made some changes so it was in a stuck state. I removed all the changes it made and ran it again with my `bundle` problems fixed so it went better this time, but it still ran into the following error.
    ```
    ➜ ./script/add-grammar https://github.com/traiansf/language-k
    npm WARN prefer global coffee-script@1.9.0 should be installed with -g
    OK vendor/grammars/language-k (source.k-framework)
    script/list-grammars:81:in `block in to_markdown': undefined method `chomp' for nil:NilClass (NoMethodError)
    	from script/list-grammars:65:in `each'
    	from script/list-grammars:65:in `to_markdown'
    	from script/list-grammars:96:in `update_readme'
    	from script/list-grammars:102:in `<main>'
    ```
    I see that a6df945c93b1771dbec4980ce97ba5284da71c39 has updated the `grammars.yml`, so this might be fine, but as I don't know all the tasks it is supposed to accomplish I can't be sure if this is enough. What should I do about this?

1. I have now updated `vendor/README.md`.

1. I just opened up an issue at `traiansf/language-k` about the permissive license issue.

1. The examples are currently under the [UIUC/NCSA license](https://opensource.org/licenses/NCSA). If you could whitelist this we would prefer that but if that's not possible, we can release these sample programs under one of the whitelisted licenses (in another repo maybe).

1. I removed most of the sample programs and left only the smaller ones. If this is too few now, I can add some of them back.

Please let me know if there is anything else I can do to help.

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -2096,6 +2096,14 @@ Jupyter Notebook:
   aliases:
   - IPython Notebook
   language_id: 185
+K Framework:
+  type: programming
+  color: "#DC143C"
```
Comment (ayberkt) [inline lib/linguist/languages.yml]: This seems to be the hardest problems :smile:. Can we try `#3E658D`?

File: lib/linguist/languages.yml [hunk 2]
```diff
@@ -2096,6 +2096,14 @@ Jupyter Notebook:
...
+  extensions:
+  - ".k"
+  aliases:
+  - K Language
+  - K
```
Comment (ayberkt) [inline lib/linguist/languages.yml]: 1a4cb4f0d8d5ee1359f1e73fc2e25f818fc6d17b

**Comment by Alhadis** [comment]
> The examples are currently under the [UIUC/NCSA license](https://opensource.org/licenses/NCSA). If you could whitelist this we would prefer that but if that's not possible, we can release these sample programs under one of the whitelisted licenses (in another repo maybe).

Paging our resident law-guy, @mlinksva.

Is [this an acceptable license](https://opensource.org/licenses/NCSA)? Can we whitelist it?

**Comment by mlinksva** [comment]
@Alhadis yes [NCSA](https://choosealicense.com/licenses/ncsa/) is acceptable and can be whitelisted.

**Comment by Alhadis** [comment]
Thanks! =) Submitted a PR to whitelist it: #3714

@ayberkt Once #3714 is merged, you might need to do a fresh pull from upstream.

**Comment by ayberkt** [comment]
Thanks @mlinksva @Alhadis!

**Comment by axic** [comment]
@ayberkt are you still pursuing this?

**Comment by ayberkt** [comment]
@axic no and I don't even remember why it wasn't concluded when I was pursuing it.

**Comment by lildude** [comment]
I know this is a really really old PR that will need a bit of work to get it ready for merging (it's probably easier to close it and open a new one), but I thought I'd point out that we now have a mechanism for adding the `.k` extension thanks to the new method of handling "generic" extensions implemented in https://github.com/github/linguist/pull/4936

In short, it involves:

- adding the `.k` extension to the list at https://github.com/github/linguist/blob/3d39c1c1a0d34c1af890ecfd18cf9ed595c68132/lib/linguist/strategy/extension.rb#L6
- adding a heuristic to clearly differentiate and identify `.k` "K Framework" files
- adding a test to confirm the clarification
- adding a sample

**Comment by ayberkt** [comment]
> I know this is a really really old PR that will need a bit of work to get it ready for merging (it's probably easier to close it and open a new one)

Thanks @lildude. I don't remember why this PR was never concluded. Did I forget to do something? In any case, I think someone from FSL would be interested in pursuing this further.

cc: @grosu @xc93 @msaxena2.

**Comment by lildude** [comment]
> I don't remember why this PR was never concluded. Did I forget to do something?

No. It was because `.k` is too generic and matches far too many files from many different languages which couldn't all be identified so the PR was blocked. We didn't have a solution for such generic cases then, but we do now 😄

**Comment by lildude** [comment]
Closing as stale. If someone wants to pick this up and implement support for the `.k` extension in future, my comment at https://github.com/github/linguist/pull/3689#issuecomment-704791687 will help you do that.

**Pull request closed without merging**